lib/puppet/provider/mongodb_replset/mongo.rb

```diff
@@ -284,13 +284,15 @@ def create_replica_set(alive_hosts)
     retry_sleep = 3
 
     retry_limit.times do |n|
-      if db_ismaster(alive_hosts[0]['host'])['ismaster']
-        Puppet.debug 'Replica set initialization has successfully ended'
-        return true
-      else
-        Puppet.debug "Waiting for replica initialization. Retry: #{n}"
-        sleep retry_sleep
-        next
+      alive_hosts.each do |alive_host|
+        if db_ismaster(alive_host['host'])['ismaster']
+          Puppet.debug 'Replica set initialization has successfully ended'
+          return true
+        else
+          Puppet.debug "Waiting for replica initialization. Retry: #{n}"
+          sleep retry_sleep
+          next
+        end
```

Comment:
I' not convinced this check actually needs to run on the master.

Reply:
The problem here is that `db_ismaster(alive_hosts[0]['host'])['ismaster']`never returns true when host[1] is elected primary, and we go to `raise Puppet::Error, "rs.initiate() failed for replicaset #{name}"`

Reply:
I mean the check to see if the replica set was successfully initiated, but I guess there should be a master somewhere...

How about this:
```
      alive_hosts.each do |alive_host|
        if db_ismaster(alive_host['host'])['ismaster']
          Puppet.debug 'Replica set initialization has successfully ended'
          return true
        end
      end
      Puppet.debug "Waiting for replica initialization. Retry: #{n}"
      sleep retry_sleep
```
This moves the sleep outside the loop over the different hosts, as it makes no sense to sleep just because the first member isn't master when the second already is.

Reply:
Your absolutely right, better to check all hosts and then sleep/retry.
Fix on it's way